File: py/src/apptheory/cache.py
from __future__ import annotations

import hashlib
from typing import Any

from apptheory.util import canonicalize_headers, to_bytes


def cache_control_ssr() -> str:
    return "private, no-store"


def cache_control_ssg() -> str:
    return "public, max-age=0, s-maxage=31536000"


def cache_control_isr(revalidate_seconds: int, stale_while_revalidate_seconds: int = 0) -> str:
    revalidate = int(revalidate_seconds or 0)
    if revalidate < 0:
        revalidate = 0
    stale = int(stale_while_revalidate_seconds or 0)
    if stale < 0:
        stale = 0

    parts = ["public", "max-age=0", f"s-maxage={revalidate}"]
    if stale > 0:
        parts.append(f"stale-while-revalidate={stale}")
    return ", ".join(parts)


def etag(body: Any) -> str:
    b = to_bytes(body)
    digest = hashlib.sha256(b).hexdigest()
    return f'"{digest}"'


def matches_if_none_match(headers: dict[str, Any] | None, etag_value: str) -> bool:
    tag = str(etag_value or "").strip()
    if not tag:
        return False

    h = canonicalize_headers(headers)
    for raw in h.get("if-none-match", []):
        for token in _split_comma_values(raw):
            if token == "*":
                return True
            token = token.strip()
            if token.lower().startswith("w/"):
                token = token[2:].strip()
            if token == tag:
                return True
    return False


def vary(existing: list[str] | None, *add: str) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []

    def add_token(token: str) -> None:
        key = str(token or "").strip().lower()
        if not key:
            return
        if key in seen:
            return
        seen.add(key)
        out.append(key)

    def add_value(value: Any) -> None:
        for token in _split_comma_values(value):
            add_token(token)

    for value in existing or []:
        add_value(value)
    for value in add:
        add_value(value)

    out.sort()
    return out


def _split_comma_values(value: Any) -> list[str]:
    raw = str(value or "").strip()
    if not raw:
        return []
    parts = [p.strip() for p in raw.split(",")]
    return [p for p in parts if p]

File: runtime/cache.go
package apptheory

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
)

// CacheControlSSR returns a conservative Cache-Control value for dynamic SSR.
func CacheControlSSR() string {
	return "private, no-store"
}

// CacheControlSSG returns a Cache-Control value suitable for edge-cached SSG HTML (no browser caching).
func CacheControlSSG() string {
	return "public, max-age=0, s-maxage=31536000"
}

// CacheControlISR returns a Cache-Control value suitable for ISR (edge-cached with stale-while-revalidate).
func CacheControlISR(revalidateSeconds, staleWhileRevalidateSeconds int) string {
	if revalidateSeconds < 0 {
		revalidateSeconds = 0
	}
	if staleWhileRevalidateSeconds < 0 {
		staleWhileRevalidateSeconds = 0
	}

	parts := []string{
		"public",
		"max-age=0",
		fmt.Sprintf("s-maxage=%d", revalidateSeconds),
	}
	if staleWhileRevalidateSeconds > 0 {
		parts = append(parts, fmt.Sprintf("stale-while-revalidate=%d", staleWhileRevalidateSeconds))
	}
	return strings.Join(parts, ", ")
}

// ETag returns a deterministic, strong ETag for the provided bytes.
//
// The returned value includes quotes (e.g. `"abc..."`).
func ETag(body []byte) string {
	sum := sha256.Sum256(body)
	return `"` + hex.EncodeToString(sum[:]) + `"`
}

// MatchesIfNoneMatch reports whether the request headers contain an If-None-Match value that matches the given etag.
func MatchesIfNoneMatch(headers map[string][]string, etag string) bool {
	etag = strings.TrimSpace(etag)
	if etag == "" {
		return false
	}

	headers = canonicalizeHeaders(headers)
	for _, header := range headers["if-none-match"] {
		for _, token := range splitCommaValues(header) {
			if token == "*" {
				return true
			}
			if strings.HasPrefix(token, "W/") {
				token = strings.TrimSpace(strings.TrimPrefix(token, "W/"))
			}
			if token == etag {
				return true
			}
		}
	}
	return false
}

// Vary merges Vary header values with deterministic ordering.
//
// Values are split on commas, trimmed, lowercased, de-duplicated, and sorted.
func Vary(existing []string, add ...string) []string {
	seen := map[string]struct{}{}
	var out []string

	addValue := func(value string) {
		value = strings.ToLower(strings.TrimSpace(value))
		if value == "" {
			return
		}
		if _, ok := seen[value]; ok {
			return
		}
		seen[value] = struct{}{}
		out = append(out, value)
	}

	for _, v := range existing {
		for _, token := range splitCommaValues(v) {
			addValue(token)
		}
	}
	for _, v := range add {
		for _, token := range splitCommaValues(v) {
			addValue(token)
		}
	}

	sort.Strings(out)
	return out
}

func splitCommaValues(value string) []string {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	parts := strings.Split(value, ",")
	out := make([]string, 0, len(parts))
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed == "" {
			continue
		}
		out = append(out, trimmed)
	}
	return out
}
